

Olá estudante!

Me chamo Rafael Meirelles e irei revisar o seu projeto hoje e em eventuais futuras submissões até que ele cumpra todos os requisitos para o aceite.
Conte comigo nessa jornada e não se preocupe se precisar ajustar alguns detalhes, é parte do processo e fundamental para que você exercite os conceitos que vem aprendendo e assim melhore a qualidade dos seus códigos e análises.

**Peço que mantenha e não altereos comentários que eu fizer por aqui para que possamos nos localizar posteriormente, ok?**

Mais uma coisa, vamos utilizar um código de cores para você entender os meus feedbacks no seu notebook. Funciona assim:


<div class="alert alert-danger">
<strong>Vermelho</strong>

Erro que precisa ser consertado, caso contrário, seu projeto não pode ser aceito

</div>

<div class="alert alert-warning">
<strong>Amarelo</strong>

Alerta de um erro não crítco, mas que pode ser corrigido para melhoria geral no seu código/análise

</div>

<div class="alert alert-success">
<strong>Verde</strong>

Elogios
</div>

<div class="alert alert-info">
<strong>Comentário do estudante</strong>

Use uma caixa azul como essa para eventuais comentários que você gostaria de fazer para mim.
</div>




<div class="alert alert-danger">
<strong>Comentário do revisor v1</strong>

Estamos no caminho certo com essa versão, mas retocar os pontos indicados abaixo nos fará chegar em um resultado ainda mais incrível.
    
Conte comigo para ajudá-lo (a) em qualquer questão que possa vir a ter! Lembre-se de usar as caixas azuis para se comunicar comigo.
    
Até breve.
</div>


<div class="alert alert-success">
<strong>Comentário do revisor v2</strong>

Trabalho muito sólido com o projeto! Gostei muito.
    
    
Parabéns pela aprovação e até breve!
</div>



# Se liga na música

# Conteúdo <a id='back'></a>

* [Introdução](#intro)
* [Etapa 1. Visão geral dos dados](#data_review)
    * [Conclusões](#data_review_conclusions)
* [Etapa 2. Pré-processamento de dados](#data_preprocessing)
    * [2.1 Estilo do cabeçalho](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusões](#data_preprocessing_conclusions)
* [Etapa 3. Teste da hipótese](#hypothesis)
    * [3.1 Hipótese 1: atividade dos usuários nas duas cidades](#activity)
* [Conclusões](#end)

## Introdução <a id='intro'></a>
O trabalho de um analista é analisar dados para obter percepções valiosas dos dados e tomar decisões fundamentadas neles. Esse processo consiste em várias etapas, como visão geral dos dados, pré-processamento dos dados e testes de hipóteses.

Sempre que fazemos uma pesquisa, precisamos formular uma hipótese que depois poderemos testar. Às vezes nós aceitamos essas hipóteses; outras vezes, nós as rejeitamos. Para fazer as escolhas certas, um negócio deve ser capaz de entender se está fazendo as suposições certas ou não.

Neste projeto, você vai comparar as preferências musicais dos habitantes de Springfild e Shelbyville. Você vai estudar os dados de um serviço de streaming de música online para testar a hipótese apresentada abaixo e comparar o comportamento dos usuários dessas duas cidades.

### Objetivo:
Teste a hipótese:
1. A atividade dos usuários é diferente dependendo do dia da semana e da cidade.


### Etapas
Os dados sobre o comportamento do usuário são armazenados no arquivo `/datasets/music_project_en.csv`. Não há informações sobre a qualidade dos dados, então será necessário examiná-los antes de testar a hipótese.

Primeiro, você avaliará a qualidade dos dados e verá se seus problemas são significativos. Depois, durante o pré-processamento dos dados, você tentará tratar dos problemas mais críticos.

O seu projeto consistirá em três etapas:
 1. Visão geral dos dados
 2. Pré-processamento de dados
 3. Teste da hipótese








[Voltar ao Índice](#back)

## Etapa 1. Visão geral dos dados <a id='data_review'></a>

Abra os dados e examine-os.

Você precisará da `pandas`, então, importe-a.

In [1]:
# importando pandas
import pandas as pd # importamos a biblioteca pandas para poder utilizar Classe DataFrame

Leia o arquivo `music_project_en.csv` da pasta `/datasets/` e salve-o na variável `df`:

In [2]:
# lendo o arquivo e armazenando em df
df = pd.read_csv("/datasets/music_project_en.csv")

Imprima as primeiras 10 linhas da tabela:

In [3]:
# obtenha as 10 primeiras 10 linhas da tabela df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtenha informações gerais sobre a tabela usando um comando. Você conhece o método para exibir informações gerais que precisamos obter.

In [4]:
# obtendo informações gerais sobre os nossos dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Aqui estão as nossas observações sobre a tabela. Ela contém sete colunas. Elas armazenam o mesmo tipo de dado: `object`.

De acordo com a documentação:
- `' userID'` — identificação do usuário
- `'Track'` — título da música
- `'artist'` — nome do artista
- `'genre'` — gênero da música
- `'City'` — cidade do usuário
- `'time'` — o tempo exato que a música foi reproduzida
- `'Day'` — dia da semana

Podemos ver três problemas de estilo nos cabeçalhos da tabela:
1. Alguns cabeçalhos são escritos em letras maiúsculas, outros estão em minúsculas.
2. Alguns cabeçalhos contêm espaços.
3. Algumas colunas possuem valores ausentes em seu conteúdo.




### Escreva suas observações. Aqui estão algumas perguntas que podem ajudar: <a id='data_review_conclusions'></a>

`1.   Que tipo de dados temos nas linhas? E como podemos entender as colunas?`
Os tipos de dados que temos em todas as colunas são Object = String, Como já sabemos os parâmetros de um DataFrame são duas listas:
1)	correspondem aos próprios dados e são listas aninhadas que são identificadas na tabela pelos seus índices, principalmente podemos definir que essas linhas contêm os dados expressos em informações organizadas para uma atividade (ouvir uma música) especifica neste caso cada usuário ou o mesmo usuário em momentos diferentes, é dizer, uma única entidade.

2)	O segundo parâmetro é outra lista que define os nomes das colunas e estabelece a estrutura para definir os atributos de cada um dos dados (entidades) embora no nosso caso sejam todos Objetc, para cada usuário são formatados como registros representados e armazenados em nossa tabela são iguais, têm peculiaridades muito precisas: por exemplo ao mesmo tempo vários usuários podem estar ouvindo a mesma música, do mesmo artista, do mesmo gênero, esta análise se aplica a todas as colunas, com base nisso analise uma restrição que devemos saber se o serviço de streaming que estamos analisando permite que o mesmo usuário reproduza música com sua conta ao mesmo tempo em dispositivos diferentes e ouvir diferentes artista e musicas:

caso sim: não teríamos certeza se as preferências refletidas não representam o usuário diretamente, mas sim um conjunto de usuários que usam essa conta.

Caso não: podemos determinar que, como o usuário só pode usar sua conta em um único dispositivo, é mais seguro que as informações reflitam com precisão as preferências do usuário.
 

`2.   Esses dados são suficientes para responder à nossa hipótese ou precisamos de mais dados?`

Embora para hipóteses estatísticas mais elaboradas faltassem parâmetros como níveis de significância alfa, erro amostral e outros parâmetros estatísticos amostrais e populacionais, para inferir sobre o comportamento dos dados deveríamos definir o tipo de distribuição probabilística de nossos dados, por exemplo, em nosso caso posso estimar que os dados são úteis e o mínimo para responder ao nosso projeto

`3.   Você notou algum problema nos dados, como valores ausentes, duplicados ou tipos de dados errados`

Observando atentamente as informações apresentadas em nossa tabela de dados, podemos ver refletidos em três colunas ('track', 'artist' e 'genre') que temos valores ausentes que são dados importantes, porque são os principais argumentos para saber as preferências dos usuários, embora no nosso caso por estarmos analisando a atividade de cada usuário, esta ausência não é muito representativa para concluir se o usuário teve atividade ou não, mas temos uma coluna que agrupa como característica das variáveis: 'track' e 'artist' que poderia ser o gênero 'genre' e ser associado como elemento comum às preferências de cada cidade.

Uma coluna muito particular é a 'hora' porque reflete a hora do dia em que a música foi tocada. O fato de ser um tipo de dados Object não faz diferença para nossa hipótese em relação à atividade do usuário, mas precisamos de informações em relação à. quantidade de tempo tocando a música, faltaria pelo menos outra coluna que representa o tempo em que o usuário parou de ouvir a música e realizar as conversões e cálculos associados para obter a quantidade de tempo

[Voltar ao Índice](#back)

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Bom trabalho com a introdução
</div>

## Etapa 2. Pré-processamento de dados <a id='data_preprocessing'></a>

O objetivo aqui é preparar os dados para a análise.
O primeiro passo é resolver todos os problemas com o cabeçalho. E então podemos passar para os valores ausentes e duplicados. Vamos começar.

Corrija a formatação nos cabeçalhos da tabela.


### Estilo do cabeçalho <a id='header_style'></a>
Imprima os cabeçalhos da tabela (os nomes das colunas):

In [5]:
print(df.columns) # imprima os nomes das colunas

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Mude os cabeçalhos da tabela conforme as boas práticas de estilo:
* Todos os caracteres precisam estar com letras minúsculas
* Exclua espaços
* Se o nome tiver várias palavras, use snake_case

Anteriormente, você aprendeu sobre uma maneira automatizada de renomear colunas. Vamos usá-la agora. Use o ciclo for para percorrer os nomes das colunas e transformar todos os caracteres em letras minúsculas. Após fazer isso, imprima os cabeçalhos da tabela novamente:

In [6]:
coluna_minuscula = []                        # Criamos uma lista vazia como auxiliar para armazenar as colunas do nosso DataFreme
for nome_antigo in  df.columns:              # Criamos um ciclo 'for' para percorrer a lista onde nossas colunas atuais estão armazenadas (df.columns) e iterar uma por uma, passando pela variável auxiliar (nome_antigo)
    nome_minuscula = nome_antigo.lower()     # Aplicamos o método lower() para converter a string de cada um para minúsculas
    coluna_minuscula.append(nome_minuscula)  # Durante cada iteração aplicamos o método append() para cada posição na lista e adicionamos os nomes modificados, se aplicável.
df.columns = coluna_minuscula                # Atribuímos a lista resultante das iterações da 'coluna minúscula' às colunas DataFrame
print(df.columns)                            # Imprimimos a lista da colunas do DataFrame atualizada

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='object')


Agora, usando a mesma abordagem, exclua os espaços no início e no final de cada nome de coluna e imprima os nomes das colunas novamente:

In [7]:
coluna_sem_espaco = []                         # Criamos uma lista vazia como auxiliar para armazenar as colunas do nosso DataFreme
for nome_antigo in  df.columns:                # Criamos um ciclo 'for' para percorrer a lista onde nossas colunas atuais estão armazenadas (df.columns) e iterar uma por uma, passando pela variável auxiliar (nome_antigo)
    nome_sem_espaço = nome_antigo.strip()      # Aplicamos o método strip() para que subtrai os espaços no início e no final da string
    coluna_sem_espaco.append(nome_sem_espaço)  # Durante cada iteração aplicamos o método append() para cada posição na lista e adicionamos os nomes modificados, se aplicável.
df.columns = coluna_sem_espaco                 # Atribuímos a lista resultante das iterações da 'coluna_sem_espaco' às colunas DataFrame
print(df.columns)                              # Imprimimos a lista da colunas do DataFrame atualizada

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Precisamos aplicar a regra de sublinhado no lugar de espaço à coluna `userid`. Deveria ser `user_id`. Renomeie essa coluna e imprima os nomes de todas as colunas quando terminar.

In [8]:
df = df.rename(columns={'userid': 'user_id'})  # Atribuímos um dicionário e alteramos o nome solicitado pelo parâmetro coluna da função rename() aplicada ao DataFrame, e atualizamos atribuindo o resultado à variável df
print(df.columns)                              # Imprimimos a lista da colunas do DataFrame atualizada

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Verifique o resultado. Imprima os cabeçalhos novamente:

In [9]:
print(df.columns)                              # Imprimimos a lista da colunas do DataFrame atualizada

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Voltar ao Índice](#back)

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Tudo certo com o nome das colunas!
</div>

### Valores Ausentes <a id='missing_values'></a>
 Primeiro, encontre a quantidade de valores ausentes na tabela. Você precisa usar dois métodos em sequência para obter o número de valores ausentes.

In [10]:
print(df.isna().sum())  # calculando o número de valores ausentes aplicando os métodos isna() e sum() 

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


Nem todos os valores ausentes afetam a pesquisa. Por exemplo, os valores ausentes em `track` e `artist` não são críticos. Você pode simplesmente substituí-los por valores padrão, como a string `'unknown'`.

Mas valores ausentes em `'genre'` podem afetar a comparação de preferências musicais de Springfield e Shelbyville. Na vida real, seria útil descobrir as razões pelas quais os dados estão ausentes e tentar corrigi-los. Mas nós não temos essa possibilidade neste projeto. Então, você terá que:
* Preencha esses valores ausentes com um valor padrão
* Avalie em que medida os valores ausentes podem afetar sua análise

Substitua os valores ausentes nas colunas `'track'`, `'artist'` e `'genre'` pela string `'unknown'`. Como mostramos nas lições anteriores, a melhor maneira de fazer isso é criar uma lista para armazenar os nomes das colunas nas quais precisamos fazer a substituição. Em seguida, use essa lista e percorra as colunas nas quais a substituição seja necessária e faça a substituição.

In [11]:
colunas_substituir = ['track', 'artist', 'genre']  # Criamos uma lista com os nomes das colunas onde temos valores ausentes para serem repetidos nos ciclos de iteração
for coluna in colunas_substituir:                  # Criamos um ciclo 'for' para percorrer a tabelas nas colunas especificadas
    df[coluna].fillna('unknown', inplace=True)     # Percorremos o DataFrame pelos índices especificados de cada coluna e aplicamos o método fillna() para substituir os valores ausentes


Agora verifique o resultado para ter certeza de que o conjunto de dados não contenha valores ausentes após a substituição. Para fazer isso, conte os valores ausentes novamente.

In [12]:
print(df.isna().sum())  # Aplicamos os métodos isna() junto com o metodo sum() para contar os valores ausentes

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Voltar ao Índice](#back)

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Valores ausentes tratados com sucesso.
</div>

### Duplicados <a id='duplicates'></a>
Encontre o número de duplicados explícitos na tabela. Lembre-se de que você precisa aplicar dois métodos em sequência para obter o número de duplicados explícitos.

In [13]:
print(df.duplicated().sum())    # contando duplicados explícitos

3826


Agora descarte todos os duplicados. Para fazer isso, chame o método que faz exatamente isso.

In [14]:
df = df.drop_duplicates().reset_index(drop=True)       # removendo duplicados explícitos

Agora vamos verificar se descartamos todos os duplicados. Conte duplicados explícitos mais uma vez para ter certeza de que você removeu todos eles:


<div class="alert alert-danger">
<strong>Comentário do revisor v1</strong>

Aqui vale a pena utilizar o .reset_index(drop=True) para garantir que não teremos problemas com a ordem das observações após removermos os duplicados:
    
Ao remover duplicatas, o índice do DataFrame pode ficar desorganizado. Isso significa que você pode acabar com lacunas no índice ou com índices que não estão mais em ordem sequencial. O método reset_index() redefine o índice do DataFrame, garantindo que ele esteja organizado novamente em uma sequência numérica contínua.
</div>



<div class="alert alert-info">
<strong>Comentário do estudante:</strong>

Absolutamente certo: esqueci que os índices originais permanecem na mesma posição e tive que atualizar os índices e eliminar a coluna com índices antigos. Aplicamos os métodos correspondentes e bem indicados para cumprir esta boa prática.

<div class="alert alert-success">
<strong>Comentário do revisor v2</strong>

Maravilha!
</div>



In [15]:
print(df.duplicated().sum())    # verificando duplicados novamente

0


Agora queremos nos livrar dos duplicados implícitos na coluna `genre`. Por exemplo, o nome de um gênero pode ser escrito de maneiras diferentes. Alguns erros afetarão também o resultado.

Para fazer isso, vamos começar imprimindo uma lista de nomes de gênero unívocos, ordenados em ordem alfabética: Para fazer isso:
* Extraia a coluna `genre` do DataFrame
* Chame o método que retornará todos os valores unívocos na coluna extraída


In [16]:
print(df['genre'].sort_values().unique()) # Usamos a indexação para extrair apenas os valores da coluna ‘gênero’, para essa coluna do DataFrame ‘df’, aplicamos o método sort_values() para ordenar a coluna e depois aplicamos o método unique() para gerar o que é requerido:uma lista de nomes de gênero unívocos, ordenados em ordem alfabética 

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Olhe a lista e encontre duplicados implícitos do gênero `hiphop`. Esses podem ser nomes escritos incorretamente, ou nomes alternativos para o mesmo gênero.

Você verá os seguintes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para se livrar deles, crie uma função `replace_wrong_genres()` com dois parâmetros:
* `wrong_genres=` — essa é uma lista que contém todos os valores que você precisa substituir
* `correct_genre=` — essa é uma string que você vai usar para a substituição

Como resultado, a função deve corrigir os nomes na coluna `'genre'` da tabela `df`, isto é, substituindo cada valor da lista `wrong_genres` por valores de `correct_genre`.

Dentro do corpo da função, use um ciclo `'for'` para percorrer a lista de gêneros errados, extrair a coluna `'genre'` e aplicar o método `replace` para fazer as correções.

In [17]:
def replace_wrong_genres(wrong_genres, correct_genre):                  # Definimos a função conforme solicitado: crie uma função replace_wrong_genres() com dois parâmetros
    for valor_errado in wrong_genres:                                   # Criamos um ciclo 'for' para que o parâmetro 'wrong_genres' (lista de valores errados) que irá receber a função percorrida em cada iteração do ciclo através da variável auxiliar de cada iteração 'valor_errado' que armazena cada valor de cada uma das posições da lista 'wrong_genres'
        df['genre'] = df['genre'].replace(valor_errado, correct_genre)  # Aplicamos a função replace() para a coluna 'genre' da DateFrame df, passamo pela função replace() os argumentos valor_errado (valor obtido da lista de valores errados de cada iteração e procurado na coluna 'genre' da DataFrame) correct_genre (valor parametro recebido dos argumentos da função quando é chamada, este é o valor pelo qual o valor errado será substituido) a aplicaçao da função replace() vai substituir valor correto pelo errado na posição onde seja localizado dentro da DataFrame          
    return df                                                           # Retorna o resultado da função que neste caso seria, devido ao DataFrame e coluna 'gênero' atualizados com valores duplicados implícitos para o gênero hiphop

duplicado_impli = ['hip', 'hip-hop', 'hop']                             # Criamos uma lista com os valores com duplicatas que pudemos observar após a classificação e posterior subtração dos valores univocos da coluna
nome_correto = 'hiphop'                                                 # Criamos uma string com o nome correcto para substituir os valores duplicados para o gênero hiphop
df = replace_wrong_genres(duplicado_impli, nome_correto)                # Chamamos a função replace_wrong_genres passando 2 argumentos que posteriormente serão usados como parâmetros da função, esses argumentos são: 'duplicate_impli' (lista de valores errados) e 'nome_correto' (nome correto para substituir na tabela) os resultados do função será atribuída ao DataFrame df

Agora, chame a função `replace_wrong_genres()` e passe argumentos apropriados para que ela limpe duplicados implícitos (`hip`, `hop` e `hip-hop`) substituindo-os por `hiphop`:

In [18]:
print(df['genre'].sort_values().unique()) # removendo duplicados implícitos, verificando se foram removidos

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

Certifique-se que os nomes duplicados foram removidos. Imprima a lista de valores unívocos da coluna `'genre'` mais uma vez:

In [19]:
print(df.duplicated().sum())  # verificando valores duplicados

0


[Voltar ao Índice](#back)

### Suas observações <a id='data_preprocessing_conclusions'></a>

` Descreva brevemente o que você reparou ao analisar duplicados, bem como a abordagem que usou para eliminá-los e os resultados que alcançou.`

Em primeiro lugar, uma prática interessante é obter os valores unívocos já ordenados porque nos permite identificar mais rapidamente os valores duplicados implícitos, ajudaria a classificá-los pela primeira letra e poder observar os valores de cada string com maior precisão, ajudaria a reduzir os grupos, por exemplo, seguindo esta logica poderíamos observar todos os gêneros agrupados cuja letra inicial é 'h', então teríamos: 'handsup' "hard'n'heavy" 'hardcore' 'hardstyle' 'hardtechno' 'hip'  'hip-hop' 'hiphop' 'historisch' 'holiday' 'hop' 'horror' 'house'  e em cada grupo detectar valores atípico como por exemplo: 'xhiphop' que estaria no grupo de valores que começam com x

Como o projeto indica abordar apenas o gênero ‘hiphop’, não podemos ter certeza se outros gêneros também possuem duplicados implícitas que poderiam distorcer as conclusões emitidas a partir das análises das informações registradas.

Outro ponto importante é que dependendo da informação que estamos analisando, sem dúvida é importante ter conhecimento a respeito dos dados que estamos recebendo, é preciso ter a clareza de saber o que vamos procurar e o que queremos como resultado, por exemplo neste caso, se quisermos saber se os géneros da nossa tabela estão corretos, deveríamos ter pelo menos uma fonte oficial para podermos comparar a nossa tabela com uma informação padrão (por exemplo: lista de todos os géneros registados a nível mundial atualmente) e não ter erros de interpretação, isso evitaria cometer por desconhecimento o seguinte erro:
    
 - Em nossa lista temos 2 gêneros que são: ‘pop’ ou ‘k-pop’ Por não termos uma fonte oficial ou uma lista confiável para comparar os diferentes tipos de gêneros registrados até o momento, podemos cometer o erro de substituir 'pop' em vez de 'k-pop' ação que nos levaria a cometer um erro na interpretação da informação.


<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Bom trabalho com os duplicados implícitos e explícitos.
</div>


[Voltar ao Índice](#back)

## Etapa 3. Teste da hipótese <a id='hypothesis'></a>

### Hipótese: comparação do comportamento dos usuários nas duas cidades <a id='activity'></a>

A hipótese afirma que existem diferenças no consumo de música pelos usuários em Springfield e em Shelbyville. Para testar a hipótese, use os dados dos três dias da semana: segunda-feira (Monday), quarta-feira (Wednesday) e sexta-feira (Friday).

* Agrupe os usuários por cidade.
* Compare o número de músicas tocadas por cada grupo na segunda, quarta e sexta.


Execute cada cálculo separadamente.

O primeiro passo é avaliar a atividade dos usuários em cada cidade. Não se esqueça das etapas "divisão-aplicação-combinação" sobre as quais falamos anteriormente na lição. Agora seu objetivo é agrupar os dados por cidade, aplicar o método de contagem apropriado durante a etapa de aplicação e então encontrar o número de músicas tocadas por cada grupo, especificando a coluna para a qual você quer obter a contagem.

Veja um exemplo de como o resultado final deve ser:
`df.groupby(by='....')['column'].method()` Execute cada cálculo separadamente.

Para avaliar a atividade dos usuários em cada cidade, agrupe os dados por cidade e encontre o número de músicas reproduzidas em cada grupo.



In [20]:
print('Dados agrupados por músicas reproduzidas em cada cidade:')
musica_cidade = df.groupby(by = 'city')['track'].count()            # Agrupamos a coluna 'city' e contamos os numeros de 'track' reproduzido por cada cidade
print()
print(musica_cidade)                  

Dados agrupados por músicas reproduzidas em cada cidade:

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


`observações:`

Agrupando os dados pelos valores das colunas podemos perceber que o número acumulado de músicas tocadas não é igual ao registro total que temos em nossa tabela (65079) e os resultados que obtivemos foram 
- Shelbyville 18512
- Springfield 42741
- total=61253 

temos 3826 valores que não estão sendo representados seria 5,88%, A partir dos resultados podemos definir que a cidade de Shelbyville teve aproximadamente 30% da atividade total registrada em nossa tabela e
Springfield 70%


Agora vamos agrupar os dados por dia da semana e encontrar a quantidade de músicas tocadas na segunda, quarta e sexta-feira. Use a mesma abordagem que antes, mas agora precisamos agrupar os dados de uma forma diferente.


In [21]:
print('Dados agrupados por músicas reproduzidas em cada dia registrado:')
musica_dia = df.groupby(by = 'day')['track'].count()                       # Agrupamos a coluna 'day' e contamos os numeros de 'track' reproduzido por cada dia
print()
print(musica_dia)

Dados agrupados por músicas reproduzidas em cada dia registrado:

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`observações:`

Podemos observar pelos resultados que a reprodução de música durante a semana (segunda, quarta, sexta) apresenta uma tendência uniforme com proporção aproximada de:

- Segunda: 21840  ---> 36%
- Quarta:   21354  ---> 35%
- Sexta:  18059  ---> 29%

Através dos resultados podemos constatar que a reprodução por dia tem uma variação baixa em termos da quantidade de música tocada nesses dias da semana de acordo com os dados registrados.


Você acabou de aprender como contar entradas agrupando-as por cidade ou por dia. E agora você precisa escrever uma função que possa contar entradas simultaneamente com base em ambos os critérios.

Crie a função `number_tracks()` para calcular o número de músicas tocadas em um determinado dia **e** em uma determinada cidade. A função deve aceitar dois parâmetros:

- `day`: um dia da semana pelo qual precisamos filtrar os dados. Por exemplo, `'Monday'`.
- `city`: uma cidade pela qual precisamos filtrar os dados. Por exemplo, `'Springfield'`.

Dentro da função, você vai aplicar uma filtragem consecutiva com indexação lógica.

Primeiro, filtre os dados por dia e então filtre a tabela resultante por cidade.

Depois de filtrar os dados usando os dois critérios, conte o número de valores na coluna 'user_id' da tabela resultante. O resultado da contagem representará o número de entradas que você quer encontrar. Armazene o resultado em uma nova variável e imprima-o.

In [22]:
def number_tracks(day, city):                                # Declare a função number_tracks() com dois parâmetros: day= e city=.
    dia = day                                                
    cidade = city 
    df_day_city = df[df['day']==dia]                         # Armazene as linhas do DataFrame em que o valor na coluna 'day' é igual ao parâmetro day=
    df_day_city = df_day_city[df_day_city['city']==cidade]   # Filtre as linhas em que o valor na coluna 'city' é igual ao parâmetro city=
    aux = df_day_city['user_id'].count()                     # Extraia a coluna 'user_id' da tabela filtrada e aplique o método count()
    return print(aux)                                        # Retorne o número dos valores da coluna 'user_id'

Chame a função `number_tracks()` seis vezes, mudando os valores dos parâmetros, para que você possa recuperar os dados de ambas as cidades para cada um dos três dias.

In [23]:
# a quantidade de músicas tocadas em Springfield na segunda-feira
number_tracks('Monday', 'Springfield')

15740


In [24]:
# a quantidade de músicas tocadas em Shelbyville na segunda-feira
number_tracks('Monday', 'Shelbyville')

5614


In [25]:
# a quantidade de músicas tocadas em Springfield na quarta-feira
number_tracks('Wednesday', 'Springfield')

11056


In [26]:
# a quantidade de músicas tocadas em Shelbyville na quarta-feira
number_tracks('Wednesday', 'Shelbyville')

7003


In [27]:
# a quantidade de músicas tocadas em Springfield na sexta-feira
number_tracks('Friday', 'Springfield')

15945


In [28]:
# a quantidade de músicas tocadas em Shelbyville na sexta-feira
number_tracks('Friday', 'Shelbyville')

5895


<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Cálculo das reproduções foi realizado com sucesso a partir da função desenvolvida.
</div>


**Conclusões**

`Comente sobre se a terceira hipótese está correta ou deve ser rejeitada. Explique seu raciocínio.`

Hipótese: comparação do comportamento dos usuários nas duas cidades = A hipótese afirma que existem diferenças no consumo de música pelos usuários em Springfield e em Shelbyville

Para emitir conclusões sobre a hipótese proposta, consolidaremos as informações resultantes na apresentação a seguir.

                  Segunda-   Quarta-   Sexta-
    cidade        feira      feira     feira   Media   Desv_Pad
    
    Springfield   15740      11056     15945   14247   2765
    Shelbyville    5614       7003      5895    6171    734

Podemos ACEITAR a hipótese

- Primeiramente vamos descrever os resultados, é importante ler atentamente o enunciado da hipótese; Não estabelece diretamente uma comparação entre o consumo dos usuários de uma cidade e outra, pelo contrário, anuncia: se existe diferença no consumo dos usuários em cada uma de suas cidades e como podemos ver, os usuários mantêm um consumo padrão com valores mais ou menos próximos aos valores médios com desvios padrão elevados.


- A decisão de aceitar que há diferença no consumo de música pelos usuários em cada uma das cidades dependerá dos limites aceitos e impostos pela empresa. Por exemplo, se estabelecermos como diferença aceitável que o desvio do consumo seja inferior a 10% do valor médio. e qualquer comportamento superior a essa quantidade de consumo do usuário será considerado diferença de consumo, ou seja, afirmar que há diferença ou não dependerá de condições estabelecidas pela empresa com base em suas necessidades como organização.


- No nosso caso, se compararmos os valores de cada uma das cidades, podemos afirmar o seguinte com os resultados obtidos com o consumo dos usuários de cada cidade:

   *Para Springfield: podemos dizer que o consumo médio é 14247, o que representa uma diferença de 29% em relação ao valor     
    mínimo (11056) e em relação ao valor mais elevado (15945) uma diferença de 10%, o que se reflecte num desvio padrão elevado 
    de 2765, que quando comparado com o valor médio expressa uma proporção de aproximadamente 20%, são valores que expressam 
    uma flutuação elevada.
    
    *Para Shelbyville: podemos dizer que o consumo médio é 6171, o que representa uma diferença de 10% em relação ao valor  
    mínimo (5614) e uma diferença de 12% em relação ao valor mais alto (7003), o que se reflecte num desvio padrão de 
    734 que quando comparado com o valor médio expressa uma proporção de aproximadamente 12% são valores que expressam uma 
    pequena oscilação.
    
   *Embora os valores de Shelbyville apresentem um consumo mais uniforme com oscilações no consumo dos usuários em torno de 10%, para a cidade de Springfield as oscilações são maiores e indicam que há diferenças de consumo entre seus usuários em dias diferentes, Revendo o enunciado da hipótese, afirma que se houver diferença de consumo entre os usuários de Springfield e Shelbyville, já sabemos que a condicional 'e' é inclusiva, portanto as 2 cidades devem ter diferenças de consumo entre seus usuários para aceite a hipótese.
   

- Finalmente, se definirmos que:
Caso as flutuações no consumo dos usuários sejam superiores a 10% nas 2 cidades, a hipótese será aceita. Caso contrário, se as oscilações forem inferiores a 10% nas 2 cidades, considera-se que não há diferença no consumo dos usuários de cada uma das cidades.
Partindo desta premissa justificamos a aceitação da hipótese de que: EXISTEM DIFERENÇAS NO CONSUMO DE MÚSICA PELOS USUÁRIOS DE SPRINGFIELD E SHELBYVILLE. porque o percentual de flutuações comparando o desvio padrão em relação à média de consumo em ambos os casos ultrapassa 10%

<div class="alert alert-success">
<strong>Comentário do revisor v1</strong>

Perfeito o trabalho com a avaliação da hipótese proposta.
    
Gostei muito do nível da discussão e profundidade da análise. Parabéns.
</div>


<div class="alert alert-info">
<strong>Comentário do estudante:</strong>

Muito obrigado Rafael. Para mim sua consideração é muito importante.

[Voltar ao Índice](#back)

# Conclusões <a id='end'></a>

`Resuma suas conclusões sobre a hipótese aqui`
- A boa formulação da hipótese e a correta interpretação são muito importantes na análise dos resultados porque a aceitação ou rejeição de uma hipótese representará a execução de uma decisão baseada na análise dos dados e informações registradas.


- O estabelecimento de regras de decisão delimitará as condições de rejeição e aceitação da hipótese e, portanto, as probabilidades de erros na tomada de decisões.


- Embora os testes de hipóteses sejam procedimentos de inferência, seus parâmetros são muito importantes, principalmente a qualidade da amostra e os erros amostrais, portanto as informações recebidas devem ser confiáveis ​​e representativas para uma estimativa eficaz.

### Importante
Em projetos de pesquisas reais, o teste estatístico de hipóteses é mais preciso e quantitativo. Observe também que conclusões sobre uma cidade inteira nem sempre podem ser tiradas a partir de dados de apenas uma fonte.

Você aprenderá mais sobre testes de hipóteses no sprint sobre a análise estatística de dados.

[Voltar ao Índice](#back)